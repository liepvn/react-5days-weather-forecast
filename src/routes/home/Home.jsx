import React from 'react';
import { Col, Container, Row } from 'react-bootstrap';

import { WeatherForecast } from './components/WeatherForecast';
import { DEFAULT_LOCATION } from './home.constant';

/**
 * Home component
 *
 * @returns {JSX.Element}
 */
const Home = () => {
  return (
    <Container>
      <Row>
        <Col lg={{ span: 6, offset: 3 }} className='pt-4'>
          <WeatherForecast defaultLocation={DEFAULT_LOCATION} />
        </Col>
      </Row>
      <Row>
        <a
          href='https://github.com/liepvn'
          target='_blank'
          rel='noopener noreferrer'
          className='text-white text-center text-decoration-none mt-3'
        >
          <i className='bi bi-github'></i> <span>Fork it on Github</span>
        </a>
      </Row>
    </Container>
  );
};

export default Home;
