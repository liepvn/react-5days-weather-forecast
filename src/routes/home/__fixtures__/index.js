export * from './home-state.factory';