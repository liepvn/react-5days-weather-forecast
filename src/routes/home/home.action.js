import { createAction } from "core/helpers";

export const ActionTypes = {
  FETCH_WEATHER_FORECAST_REQUEST: 'FETCH_WEATHER_FORECAST_REQUEST',
  FETCH_WEATHER_FORECAST_SUCCESS: 'FETCH_WEATHER_FORECAST_SUCCESS',
  FETCH_WEATHER_FORECAST_FAILURE: 'FETCH_WEATHER_FORECAST_FAILURE',
  SEARCH_LOCATIONS_REQUEST: 'SEARCH_LOCATIONS_REQUEST',
  SEARCH_LOCATIONS_SUCCESS: 'SEARCH_LOCATIONS_SUCCESS',
  SEARCH_LOCATIONS_FAILURE: 'SEARCH_LOCATIONS_FAILURE',
}

export const fetchWeatherForecast = createAction(ActionTypes.FETCH_WEATHER_FORECAST_REQUEST, ({ woeid }) => ({ woeid }));
export const searchLocations = createAction(ActionTypes.SEARCH_LOCATIONS_REQUEST, (query) => ({ query }));