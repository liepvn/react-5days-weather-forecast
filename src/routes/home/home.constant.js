export const WEATHER_STATE_IMAGE_BASE_URL = 'https://www.metaweather.com/static/img/weather/png/64';

export const DEFAULT_LOCATION = Object.freeze({ woeid: 1252431, title: 'Ho Chi Minh City' });