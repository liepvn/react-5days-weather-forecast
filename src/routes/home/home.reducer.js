import { createReducer } from 'core/helpers';
import { take } from 'lodash';

import { ActionTypes } from './home.action';
import { WEATHER_STATE_IMAGE_BASE_URL } from './home.constant';

/**
 * Initial state
 */
export const homeState = {
  data: undefined,
  isSearching: false,
  isLoading: false,
  locations: [],
  forecast: []
};

/**
 * Reducer
 */
const homeReducer = createReducer(
  {
    [ActionTypes.FETCH_WEATHER_FORECAST_REQUEST]: (draft) => {
      draft.isLoading = true
    },
    [ActionTypes.FETCH_WEATHER_FORECAST_SUCCESS]: (draft, { payload }) => {
      const { consolidated_weather } = payload;

      draft.forecast = take(consolidated_weather, 5).map(
        ({ max_temp, min_temp, applicable_date, weather_state_abbr, humidity, the_temp }) => ({
          maxTemp: max_temp,
          minTemp: min_temp,
          date: applicable_date,
          weatherStateAbbr: weather_state_abbr,
          weatherStateImage: `${WEATHER_STATE_IMAGE_BASE_URL}/${weather_state_abbr}.png`,
          humidity,
          temp: the_temp
        })
      );
      draft.isLoading = false
    },
    [ActionTypes.FETCH_WEATHER_FORECAST_FAILURE]: (draft) => {
      draft.isLoading = false
    },
    [ActionTypes.SEARCH_LOCATIONS_REQUEST]: (draft) => {
      draft.isSearching = true
    },
    [ActionTypes.SEARCH_LOCATIONS_SUCCESS]: (draft, { payload }) => {
      draft.isSearching = false
      draft.locations = payload
    },
    [ActionTypes.SEARCH_LOCATIONS_FAILURE]: (draft) => {
      draft.isSearching = false
    },
  },
  homeState
);

const reducer = {
  home: homeReducer
}

export default reducer;
