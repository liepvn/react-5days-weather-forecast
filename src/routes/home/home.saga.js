import { all, call, put, takeLatest } from 'redux-saga/effects';
import weatherApi from 'shared/apis/weather.api';

import { ActionTypes } from './home.action';

/**
 * Handle fetching weather forecast data
 */
export function* fetchWeatherForecast({ payload }) {
  const { woeid } = payload || {};

  try {
    const { data } = yield call(weatherApi.fetchForecastByLocation, woeid);

    yield put({
      type: ActionTypes.FETCH_WEATHER_FORECAST_SUCCESS,
      payload: data
    });
  } catch (error) {
    yield put({
      type: ActionTypes.FETCH_WEATHER_FORECAST_FAILURE,
      payload: error
    });
  }
}

/**
 * Handle search locations
 */
export function* searchLocations({ payload: { query }}) {
  try {
    const { data } = yield call(weatherApi.searchLocations, query);

    yield put({
      type: ActionTypes.SEARCH_LOCATIONS_SUCCESS,
      payload: data
    });
  } catch (error) {
    yield put({
      type: ActionTypes.SEARCH_LOCATIONS_FAILURE,
      payload: error
    });
  }
}

export default function* root() {
  yield all([
    takeLatest(ActionTypes.FETCH_WEATHER_FORECAST_REQUEST, fetchWeatherForecast),
    takeLatest(ActionTypes.SEARCH_LOCATIONS_REQUEST, searchLocations)
  ]);
}
