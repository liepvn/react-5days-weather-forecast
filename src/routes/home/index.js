import Home from './Home';

import homeReducer from './home.reducer';
import homeSaga from './home.saga';

export { homeReducer, homeSaga };

export default Home;
