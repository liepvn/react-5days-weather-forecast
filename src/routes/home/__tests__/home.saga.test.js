import { expectSaga } from 'redux-saga-test-plan';

import home, { fetchWeatherForecast, searchLocations } from 'routes/home/home.saga';
import sanLocations from 'shared/apis/__fixtures__/weather-api-location-search-san.json';
import sanForecast from 'shared/apis/__fixtures__/weather-api-location-2487956.json';

import { homeReducer } from '..';
import { ActionTypes } from '../home.action';

const mockSearchLocationsApi = jest.fn();
const mockFetchForecastByLocationApi = jest.fn();
jest.mock('shared/apis/weather.api', () => ({
  searchLocations: (...args) => mockSearchLocationsApi(...args),
  fetchForecastByLocation: (...args) => mockFetchForecastByLocationApi(...args),
}));

describe('home.saga', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should have the expected watchers', async () => {
    const result = await expectSaga(home)
      .run({ silenceTimeout: true });
    expect(result.toJSON()).toMatchSnapshot();
  });

  describe('searchLocations', () => {
    const requestAction = {
      type: ActionTypes.SEARCH_LOCATIONS_REQUEST,
      payload: { query: 'san' },
    };

    it('should handle SUCCESS', async () => {
      mockSearchLocationsApi.mockResolvedValueOnce({ data: sanLocations });

      const result = await expectSaga(searchLocations, requestAction)
        .withReducer(homeReducer.home, homeReducer.home(undefined, requestAction))
        .run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(mockSearchLocationsApi).toHaveBeenCalled();
    });

    it('should handle FAILURE', async () => {
      mockSearchLocationsApi.mockRejectedValueOnce(new Error('Failed to fetch'));

      const result = await expectSaga(searchLocations, {
        type: ActionTypes.SEARCH_LOCATIONS_REQUEST,
        payload: { query: 'san' },
      })
      .withReducer(homeReducer.home, homeReducer.home(undefined, requestAction))
      .run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(mockSearchLocationsApi).toHaveBeenCalled();
    });
  });

  describe('fetchWeatherForecast', () => {
    const requestAction = {
      type: ActionTypes.FETCH_WEATHER_FORECAST_REQUEST,
      payload: { woeid: 2122 },
    };

    it('should handle SUCCESS', async () => {
      mockFetchForecastByLocationApi.mockResolvedValueOnce({ data: sanForecast });

      const result = await expectSaga(fetchWeatherForecast, requestAction)
        .withReducer(homeReducer.home, homeReducer.home(undefined, requestAction))
        .run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(mockFetchForecastByLocationApi).toHaveBeenCalled();
    });

    it('should handle FAILURE', async () => {
      mockFetchForecastByLocationApi.mockRejectedValueOnce(new Error('Failed to fetch'));

      const result = await expectSaga(fetchWeatherForecast, requestAction)
      .withReducer(homeReducer.home, homeReducer.home(undefined, requestAction))
      .run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(mockFetchForecastByLocationApi).toHaveBeenCalled();
    });
  });
});
