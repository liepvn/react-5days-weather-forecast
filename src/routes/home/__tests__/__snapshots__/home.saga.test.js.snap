// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home.saga fetchWeatherForecast should handle FAILURE 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          2122,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchForecastByLocation",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {},
          "type": "FETCH_WEATHER_FORECAST_FAILURE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`home.saga fetchWeatherForecast should handle SUCCESS 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          2122,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchForecastByLocation",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "consolidated_weather": Array [
              Object {
                "air_pressure": 1015,
                "applicable_date": "2021-08-15",
                "created": "2021-08-15T15:56:17.241083Z",
                "humidity": 74,
                "id": 6301494061039616,
                "max_temp": 19.314999999999998,
                "min_temp": 14.5,
                "predictability": 71,
                "the_temp": 21.310000000000002,
                "visibility": 14.458686272170524,
                "weather_state_abbr": "hc",
                "weather_state_name": "Heavy Cloud",
                "wind_direction": 254.6399627762891,
                "wind_direction_compass": "WSW",
                "wind_speed": 6.751997544887192,
              },
              Object {
                "air_pressure": 1009.5,
                "applicable_date": "2021-08-16",
                "created": "2021-08-15T15:56:19.970425Z",
                "humidity": 66,
                "id": 6710658717450240,
                "max_temp": 21.21,
                "min_temp": 14.68,
                "predictability": 70,
                "the_temp": 22.205,
                "visibility": 15.356567644953472,
                "weather_state_abbr": "lc",
                "weather_state_name": "Light Cloud",
                "wind_direction": 269.0369380717442,
                "wind_direction_compass": "W",
                "wind_speed": 8.885437476126091,
              },
              Object {
                "air_pressure": 1009,
                "applicable_date": "2021-08-17",
                "created": "2021-08-15T15:56:22.969805Z",
                "humidity": 72,
                "id": 5285622894821376,
                "max_temp": 18.375,
                "min_temp": 14.254999999999999,
                "predictability": 70,
                "the_temp": 20.46,
                "visibility": 15.31524646066969,
                "weather_state_abbr": "lc",
                "weather_state_name": "Light Cloud",
                "wind_direction": 259.38349247255644,
                "wind_direction_compass": "W",
                "wind_speed": 8.407470930623823,
              },
              Object {
                "air_pressure": 1011.5,
                "applicable_date": "2021-08-18",
                "created": "2021-08-15T15:56:26.048771Z",
                "humidity": 65,
                "id": 6411522801664000,
                "max_temp": 20.41,
                "min_temp": 13.754999999999999,
                "predictability": 70,
                "the_temp": 21.509999999999998,
                "visibility": 14.195224886661894,
                "weather_state_abbr": "lc",
                "weather_state_name": "Light Cloud",
                "wind_direction": 224.4286329050097,
                "wind_direction_compass": "SW",
                "wind_speed": 6.683667137031736,
              },
              Object {
                "air_pressure": 1013.5,
                "applicable_date": "2021-08-19",
                "created": "2021-08-15T15:56:28.975704Z",
                "humidity": 70,
                "id": 4564412020031488,
                "max_temp": 20.205,
                "min_temp": 14.52,
                "predictability": 68,
                "the_temp": 20.865000000000002,
                "visibility": 14.28936262228585,
                "weather_state_abbr": "c",
                "weather_state_name": "Clear",
                "wind_direction": 231.04043295874322,
                "wind_direction_compass": "SW",
                "wind_speed": 6.865788047167211,
              },
              Object {
                "air_pressure": 1011,
                "applicable_date": "2021-08-20",
                "created": "2021-08-15T15:56:32.043415Z",
                "humidity": 74,
                "id": 5177095178158080,
                "max_temp": 20.085,
                "min_temp": 14.469999999999999,
                "predictability": 70,
                "the_temp": 19.89,
                "visibility": 9.999726596675416,
                "weather_state_abbr": "lc",
                "weather_state_name": "Light Cloud",
                "wind_direction": 246.5,
                "wind_direction_compass": "WSW",
                "wind_speed": 5.507510737294202,
              },
            ],
            "latt_long": "37.777119, -122.41964",
            "location_type": "City",
            "parent": Object {
              "latt_long": "37.271881,-119.270233",
              "location_type": "Region / State / Province",
              "title": "California",
              "woeid": 2347563,
            },
            "sources": Array [
              Object {
                "crawl_rate": 360,
                "slug": "bbc",
                "title": "BBC",
                "url": "http://www.bbc.co.uk/weather/",
              },
              Object {
                "crawl_rate": 480,
                "slug": "forecast-io",
                "title": "Forecast.io",
                "url": "http://forecast.io/",
              },
              Object {
                "crawl_rate": 360,
                "slug": "hamweather",
                "title": "HAMweather",
                "url": "http://www.hamweather.com/",
              },
              Object {
                "crawl_rate": 180,
                "slug": "met-office",
                "title": "Met Office",
                "url": "http://www.metoffice.gov.uk/",
              },
              Object {
                "crawl_rate": 360,
                "slug": "openweathermap",
                "title": "OpenWeatherMap",
                "url": "http://openweathermap.org/",
              },
              Object {
                "crawl_rate": 720,
                "slug": "wunderground",
                "title": "Weather Underground",
                "url": "https://www.wunderground.com/?apiref=fc30dc3cd224e19b",
              },
              Object {
                "crawl_rate": 360,
                "slug": "world-weather-online",
                "title": "World Weather Online",
                "url": "http://www.worldweatheronline.com/",
              },
            ],
            "sun_rise": "2021-08-15T06:25:33.893965-07:00",
            "sun_set": "2021-08-15T20:03:13.312463-07:00",
            "time": "2021-08-15T10:37:34.407402-07:00",
            "timezone": "US/Pacific",
            "timezone_name": "LMT",
            "title": "San Francisco",
            "woeid": 2487956,
          },
          "type": "FETCH_WEATHER_FORECAST_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`home.saga searchLocations should handle FAILURE 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "san",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/searchLocations",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {},
          "type": "SEARCH_LOCATIONS_FAILURE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`home.saga searchLocations should handle SUCCESS 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "san",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/searchLocations",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Array [
            Object {
              "latt_long": "37.777119, -122.41964",
              "location_type": "City",
              "title": "San Francisco",
              "woeid": 2487956,
            },
            Object {
              "latt_long": "32.715691,-117.161720",
              "location_type": "City",
              "title": "San Diego",
              "woeid": 2487889,
            },
            Object {
              "latt_long": "37.338581,-121.885567",
              "location_type": "City",
              "title": "San Jose",
              "woeid": 2488042,
            },
            Object {
              "latt_long": "29.424580,-98.494614",
              "location_type": "City",
              "title": "San Antonio",
              "woeid": 2487796,
            },
            Object {
              "latt_long": "36.974018,-122.030952",
              "location_type": "City",
              "title": "Santa Cruz",
              "woeid": 2488853,
            },
            Object {
              "latt_long": "-33.463039,-70.647942",
              "location_type": "City",
              "title": "Santiago",
              "woeid": 349859,
            },
            Object {
              "latt_long": "36.406651,25.456530",
              "location_type": "City",
              "title": "Santorini",
              "woeid": 56558361,
            },
            Object {
              "latt_long": "43.461498,-3.810010",
              "location_type": "City",
              "title": "Santander",
              "woeid": 773964,
            },
            Object {
              "latt_long": "35.170429,128.999481",
              "location_type": "City",
              "title": "Busan",
              "woeid": 1132447,
            },
            Object {
              "latt_long": "28.46163,-16.267059",
              "location_type": "City",
              "title": "Santa Cruz de Tenerife",
              "woeid": 773692,
            },
            Object {
              "latt_long": "35.666431,-105.972572",
              "location_type": "City",
              "title": "Santa Fe",
              "woeid": 2488867,
            },
          ],
          "type": "SEARCH_LOCATIONS_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`home.saga should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "FETCH_WEATHER_FORECAST_REQUEST",
          "@@redux-saga-test-plan/json/function/fetchWeatherForecast",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "SEARCH_LOCATIONS_REQUEST",
          "@@redux-saga-test-plan/json/function/searchLocations",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "FETCH_WEATHER_FORECAST_REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "SEARCH_LOCATIONS_REQUEST",
      },
      "type": "TAKE",
    },
  ],
}
`;
